#module load GCCcore/.8.2.0
#module load Metal

#metal /data/davis_lab/golevasb/Projects/PNES_PRS/GWAS/Meta_Analysis_cleaned_ss/metal_script.sh > /data/davis_lab/golevasb/Projects/PNES_PRS/GWAS/Meta_Analysis_cleaned_ss/metal_fs_gwas_try.log


CUSTOMVARIABLE N

GENOMICCONTROL OFF

AVERAGEFREQ ON
MINMAXFREQ ON

SCHEME SAMPLESIZE


###if you are using sumstats generated by the pre_meta_analysis_qc.R pipeline, then these labels should be standard
SEPARATOR TAB
MARKER RS_ID
ALLELE EA OA
PVALUELABEL P
EFFECTLABEL BETA #change to log(OR) if OR instead of BETA
WEIGHTLABEL N #if no N column, remove and uncomment out two lines beneath this
#WEIGHTLABEL DONTUSECOLUMN
#DEFAULTWEIGHT 10000
STDERR SE
FREQLABEL EAF


###add each sumstats file to be processed. The output file from the pre_meta R script should go here
#in this example, I am processing 6 different summary statistics
PROCESS /data/davis_lab/golevasb/Projects/PNES_PRS/GWAS/Meta_Analysis_cleaned_ss/iPSYCH/iPSYCH_autosome_sumstats_for_meta_analysis_cleaned.txt

PROCESS /data/davis_lab/golevasb/Projects/PNES_PRS/GWAS/Meta_Analysis_cleaned_ss/MGBB/MGBB_autosome_sumstats_for_meta_analysis_cleaned.txt

PROCESS /data/davis_lab/golevasb/Projects/PNES_PRS/GWAS/Meta_Analysis_cleaned_ss/BioVU/BioVU_autosome_sumstats_for_meta_analysis_cleaned.txt

PROCESS /data/davis_lab/golevasb/Projects/PNES_PRS/GWAS/Meta_Analysis_cleaned_ss/CC/CC_autosome_sumstats_for_meta_analysis_cleaned.txt

PROCESS /data/davis_lab/golevasb/Projects/PNES_PRS/GWAS/Meta_Analysis_cleaned_ss/BioMe/BioMe_autosome_sumstats_for_meta_analysis_cleaned.txt

PROCESS /data/davis_lab/golevasb/Projects/PNES_PRS/GWAS/Meta_Analysis_cleaned_ss/MVP/MVP_autosome_sumstats_for_meta_analysis_cleaned.txt


#change out file, format is "directory/file_name .tbl"
OUTFILE /data/davis_lab/golevasb/Projects/PNES_PRS/GWAS/Meta_Analysis_cleaned_ss/metal_fs_gwas_try .tbl
ANALYZE HETEROGENEITY

#file will be written out as directory/file_name1.tbl
#log file will be written out to directory/file_name.log as specified in the 3rd commented line after the >
