module load GCCcore/.8.2.0
module load Metal

metal

CUSTOMVARIABLE TotalSampleSize
LABEL TotalSampleSize as N


GENOMICCONTROL OFF

AVERAGEFREQ ON
MINMAXFREQ ON

SCHEME SAMPLESIZE


###if you are using sumstats generated by the pre_meta_analysis_qc.R pipeline, then these labels should be standard
SEPARATOR TAB
MARKER chr_pos
ALLELE EA OA
PVALUELABEL P
EFFECTLABEL BETA #change to log(OR) if OR instead of BETA
WEIGHTLABEL N #if no N column, remove and uncomment out two lines beneath this
#WEIGHTLABEL DONTUSECOLUMN
#DEFAULTWEIGHT 10000
STDERR SE
FREQLABEL EAF


###add each sumstats file to be processed. The output file from the pre_meta R script should go here
#in this example, I am processing 6 different summary statistics
PROCESS /data/davis_lab/golevasb/Projects/PNES_PRS/GWAS/Meta_Analysis_cleaned_ss_absolute_filter_ref_maf/iPSYCH/iPSYCH_autosome_sumstats_for_meta_analysis_cleaned.txt

PROCESS /data/davis_lab/golevasb/Projects/PNES_PRS/GWAS/Meta_Analysis_cleaned_ss_absolute_filter_ref_maf/MGBB/MGBB_autosome_sumstats_for_meta_analysis_cleaned.txt

PROCESS /data/davis_lab/golevasb/Projects/PNES_PRS/GWAS/Meta_Analysis_cleaned_ss_absolute_filter_ref_maf/BioVU/BioVU_autosome_sumstats_for_meta_analysis_cleaned.txt

PROCESS /data/davis_lab/golevasb/Projects/PNES_PRS/GWAS/Meta_Analysis_cleaned_ss_absolute_filter_ref_maf/CC/CC_autosome_sumstats_for_meta_analysis_cleaned.txt

PROCESS /data/davis_lab/golevasb/Projects/PNES_PRS/GWAS/Meta_Analysis_cleaned_ss_absolute_filter_ref_maf/BioMe/BioMe_autosome_sumstats_for_meta_analysis_cleaned.txt

PROCESS /data/davis_lab/golevasb/Projects/PNES_PRS/GWAS/Meta_Analysis_cleaned_ss_absolute_filter_ref_maf/MVP/MVP_autosome_sumstats_for_meta_analysis_cleaned.txt


## WARNING: Bad alleles for marker '2:148178838', expecting 'a/g' found 'a/t'
## WARNING: Bad alleles for marker '9:120860690', expecting 'a/t' found 'a/g'


#change out file, format is "directory/file_name .tbl"
OUTFILE /data/davis_lab/golevasb/Projects/PNES_PRS/GWAS/Meta_Analysis_cleaned_ss_absolute_filter_ref_maf/metal_fs_gwas_ .tbl
ANALYZE HETEROGENEITY > metal_file_name.log

#file will be written out as directory/file_name1.tbl
